File: macos/Cleer/DSP/FeedbackSuppressor.swift
import Foundation

/// Adaptive feedback (howl) suppression — the Swift port of the validated
/// `feedback_suppressor.py`. Detection is frequency-domain (peaks that are
/// tonal, prominent, in-band and persistent); the actuator is a pool of
/// time-domain biquad notches whose depth tracks each howl.
///
/// See the Python reference for the full commentary on the algorithm.
final class FeedbackSuppressor {
    struct Params {
        var fLo = 150.0, fHi = 9000.0
        var prominenceDB = 12.0
        var holdDB = 6.0
        var persistFrames = 6.0
        var maxDepthDB = 36.0
        var q = 18.0
        var attackDB = 12.0
        var releaseDB = 1.5
        var maxNotches = 16
        var floorMedian = 41   // bins for the spectral-floor median filter
    }

    private final class Notch {
        let biquad: Biquad
        var freq: Double
        var depthDB = 0.0
        var idle = 0
        init(biquad: Biquad, freq: Double) { self.biquad = biquad; self.freq = freq }
    }

    let sampleRate: Double
    let win: Int
    let hop: Int
    var params: Params
    var enabled = true

    private let analyzer: SpectrumAnalyzer
    private let bins: Int
    private var persist: [Double]
    private var notches: [Notch] = []
    private(set) var detectedFrequencies: [Double] = []

    init(sampleRate: Double, win: Int, hop: Int, params: Params = Params()) {
        self.sampleRate = sampleRate; self.win = win; self.hop = hop
        self.params = params
        self.analyzer = SpectrumAnalyzer(win: win, hop: hop)
        self.bins = win / 2 + 1
        self.persist = [Double](repeating: 0, count: bins)
    }

    private func binToFreq(_ k: Int) -> Double { Double(k) * sampleRate / Double(win) }
    private func freqToBin(_ f: Double) -> Int { Int((f * Double(win) / sampleRate).rounded()) }

    /// Process one hop-sized block in place (time domain) and update notches.
    func process(_ block: inout [Double]) {
        let mags = analyzer.push(block)
        if enabled { updateNotches(mags) }
        guard enabled, !notches.isEmpty else { return }
        for i in 0..<block.count {
            var s = block[i]
            for n in notches { s = n.biquad.process(s) }
            block[i] = s
        }
    }

    private func updateNotches(_ mags: [Double]) {
        // Power and spectral floor (median over frequency).
        var power = [Double](repeating: 0, count: bins)
        for k in 0..<bins { power[k] = mags[k] * mags[k] + 1e-12 }
        let floor = medianFilter(power, radius: params.floorMedian / 2)

        var promDB = [Double](repeating: 0, count: bins)
        for k in 0..<bins { promDB[k] = 10.0 * log10(power[k] / (floor[k] + 1e-12)) }

        // Candidate peaks: in-band tonal local maxima above the floor.
        var candidate = [Bool](repeating: false, count: bins)
        for k in 1..<(bins - 1) {
            let f = binToFreq(k)
            if f >= params.fLo && f <= params.fHi
                && power[k] > power[k - 1] && power[k] > power[k + 1]
                && promDB[k] > params.prominenceDB {
                candidate[k] = true
            }
        }

        // Persistence: tonal feedback survives, transients decay.
        for k in 0..<bins { persist[k] = candidate[k] ? persist[k] + 1.0 : persist[k] * 0.5 }

        // Allocate notches for confirmed howls.
        for k in 0..<bins where persist[k] >= params.persistFrames {
            let already = notches.contains { abs(freqToBin($0.freq) - k) <= 1 }
            if already || notches.count >= params.maxNotches { continue }
            let freq = interpFreq(power, k)
            let n = Notch(biquad: Biquad(sampleRate: sampleRate), freq: freq)
            notches.append(n)
            detectedFrequencies.append(freq)
            if detectedFrequencies.count > 64 { detectedFrequencies.removeFirst() }
        }

        // Drive depths; release & reap dead notches.
        var survivors: [Notch] = []
        for n in notches {
            let b = freqToBin(n.freq)
            let hot = b >= 0 && b < bins && promDB[b] > params.holdDB
            let target = hot ? -params.maxDepthDB : 0.0
            if n.depthDB > target { n.depthDB = max(target, n.depthDB - params.attackDB) }
            else { n.depthDB = min(target, n.depthDB + params.releaseDB) }
            n.biquad.setPeaking(freq: n.freq, q: params.q, gainDB: n.depthDB)
            n.idle = n.depthDB >= -1e-6 ? n.idle + 1 : 0
            if n.idle < 20 { survivors.append(n) }
        }
        notches = survivors
    }

    private func interpFreq(_ power: [Double], _ k: Int) -> Double {
        guard k > 0, k < bins - 1 else { return binToFreq(k) }
        let a = log(power[k - 1] + 1e-12), b = log(power[k] + 1e-12), c = log(power[k + 1] + 1e-12)
        let denom = a - 2 * b + c
        var delta = denom != 0 ? 0.5 * (a - c) / denom : 0.0
        delta = min(0.5, max(-0.5, delta))
        return (Double(k) + delta) * sampleRate / Double(win)
    }

    var activeNotchCount: Int { notches.filter { $0.depthDB < -1.0 }.count }
}

/// Simple sliding median over frequency — estimates the broadband spectral
/// floor so that narrow tonal peaks stand out from it.
func medianFilter(_ x: [Double], radius: Int) -> [Double] {
    let n = x.count
    var out = [Double](repeating: 0, count: n)
    var windowBuf = [Double]()
    windowBuf.reserveCapacity(2 * radius + 1)
    for i in 0..<n {
        windowBuf.removeAll(keepingCapacity: true)
        let lo = max(0, i - radius), hi = min(n - 1, i + radius)
        for j in lo...hi { windowBuf.append(x[j]) }
        windowBuf.sort()
        out[i] = windowBuf[windowBuf.count / 2]
    }
    return out
}

File: reference/defeedback/feedback_suppressor.py
"""Adaptive feedback (howl) suppression.

This is the defining feature of a "de-feedback" tool. Microphone feedback
is a closed acoustic loop (speaker -> mic -> amp -> speaker) that builds up
energy at the room/mic resonance whose loop gain exceeds unity. In the
spectrum it shows up as a *narrow, tonal, persistent and rising* peak — a
"howl" or "ring" — which is exactly what distinguishes it from voice or
music (which are broadband and time-varying).

Algorithm (the same idea behind dbx AFS, Behringer FBQ, etc.):

  1.  Detector (frequency domain): every hop, FFT the latest window, find
      spectral peaks that stick far above the local spectral floor, sit in
      the feedback-prone band, and *persist* across consecutive frames.
  2.  Actuator (time domain): for each confirmed howl, place an adaptive
      parametric notch (a biquad) at the precise frequency. Depth grows
      fast while the howl is present (attack) and is released slowly once
      it is gone, then the slot is freed for re-use.

Detection is frequency-domain (cheap, accurate); filtering is time-domain
(zero added latency — no overlap-add delay). That split is why these
suppressors can run "with essentially zero latency".
"""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np
from scipy.ndimage import median_filter
from scipy.signal import lfilter, lfilter_zi

from .stft import hann


def peaking_biquad(sr: float, f0: float, q: float, gain_db: float):
    """RBJ cookbook peaking-EQ biquad. Negative gain_db => a notch."""
    A = 10.0 ** (gain_db / 40.0)
    w0 = 2.0 * np.pi * f0 / sr
    cosw = np.cos(w0)
    alpha = np.sin(w0) / (2.0 * q)
    b0 = 1.0 + alpha * A
    b1 = -2.0 * cosw
    b2 = 1.0 - alpha * A
    a0 = 1.0 + alpha / A
    a1 = -2.0 * cosw
    a2 = 1.0 - alpha / A
    b = np.array([b0, b1, b2]) / a0
    a = np.array([1.0, a1 / a0, a2 / a0])
    return b, a


@dataclass
class Notch:
    freq: float
    q: float
    depth_db: float = 0.0          # current depth (<= 0)
    target_db: float = 0.0         # where depth is heading
    zi: np.ndarray = field(default=None)  # biquad state for sample continuity
    b: np.ndarray = field(default=None)
    a: np.ndarray = field(default=None)
    idle: int = 0                  # frames spent fully released


class FeedbackSuppressor:
    def __init__(
        self,
        sr: float,
        win: int = 1024,
        hop: int = 256,
        max_notches: int = 16,
        f_lo: float = 150.0,
        f_hi: float = 9000.0,
        prominence_db: float = 12.0,   # how far above the floor counts as a peak
        hold_db: float = 6.0,          # keep a notch while peak stays this high
        persist_frames: int = 6,       # frames a peak must survive to be "feedback"
        max_depth_db: float = 36.0,    # deepest notch
        q: float = 18.0,               # notch sharpness
        attack_db: float = 12.0,       # depth added per frame while howling
        release_db: float = 1.5,       # depth removed per frame once gone
    ):
        self.sr = sr
        self.win = win
        self.hop = hop
        self.window = hann(win)
        self.max_notches = max_notches
        self.f_lo, self.f_hi = f_lo, f_hi
        self.prominence_db = prominence_db
        self.hold_db = hold_db
        self.persist_frames = persist_frames
        self.max_depth_db = max_depth_db
        self.q = q
        self.attack_db = attack_db
        self.release_db = release_db

        self.bins = win // 2 + 1
        self.fft_freqs = np.fft.rfftfreq(win, d=1.0 / sr)
        self.persist = np.zeros(self.bins)     # per-bin persistence counter
        self.notches: list[Notch] = []
        self._log: list[float] = []            # freqs we placed notches at

    # ---- detection -------------------------------------------------------
    def _detect(self, ring: np.ndarray) -> list[tuple[float, int]]:
        """Return [(freq_hz, bin)] of bins currently confirmed as feedback."""
        mag = np.abs(np.fft.rfft(ring * self.window)) + 1e-12
        power = mag ** 2
        # Spectral floor: wide median over frequency = the broadband content.
        floor = median_filter(power, size=41, mode="nearest") + 1e-12
        prom_db = 10.0 * np.log10(power / floor)

        in_band = (self.fft_freqs >= self.f_lo) & (self.fft_freqs <= self.f_hi)
        is_local_max = np.zeros(self.bins, dtype=bool)
        is_local_max[1:-1] = (power[1:-1] > power[:-2]) & (power[1:-1] > power[2:])
        candidate = in_band & is_local_max & (prom_db > self.prominence_db)

        # Persistence: tonal peaks survive; transients decay away.
        self.persist[candidate] += 1.0
        self.persist[~candidate] *= 0.5

        confirmed = []
        for k in np.nonzero(self.persist >= self.persist_frames)[0]:
            confirmed.append((self._interp_freq(power, k), int(k)))
        return confirmed

    def _interp_freq(self, power: np.ndarray, k: int) -> float:
        """Parabolic interpolation around bin k for sub-bin centre frequency."""
        if 0 < k < self.bins - 1:
            a, b, c = (np.log(power[k - 1] + 1e-12),
                       np.log(power[k] + 1e-12),
                       np.log(power[k + 1] + 1e-12))
            denom = (a - 2 * b + c)
            delta = 0.5 * (a - c) / denom if denom != 0 else 0.0
            delta = float(np.clip(delta, -0.5, 0.5))
        else:
            delta = 0.0
        return (k + delta) * self.sr / self.win

    # ---- notch pool management ------------------------------------------
    def _bin_of(self, freq: float) -> int:
        return int(round(freq * self.win / self.sr))

    def _update_notches(self, confirmed: list[tuple[float, int]], ring: np.ndarray):
        mag = np.abs(np.fft.rfft(ring * self.window)) + 1e-12
        power = mag ** 2
        floor = median_filter(power, size=41, mode="nearest") + 1e-12
        prom_db = 10.0 * np.log10(power / floor)

        confirmed_bins = {b for _, b in confirmed}
        active_bins = {self._bin_of(n.freq) for n in self.notches}

        # Allocate notches for newly-confirmed howls.
        for freq, b in confirmed:
            if any(abs(b - self._bin_of(n.freq)) <= 1 for n in self.notches):
                continue  # already covered
            if len(self.notches) >= self.max_notches:
                break
            n = Notch(freq=freq, q=self.q, target_db=-self.max_depth_db)
            self.notches.append(n)
            self._log.append(freq)

        # Drive each notch's depth and release dead ones.
        survivors = []
        for n in self.notches:
            b = self._bin_of(n.freq)
            still_hot = (0 <= b < self.bins) and (prom_db[b] > self.hold_db)
            n.target_db = -self.max_depth_db if still_hot else 0.0
            if n.depth_db > n.target_db:
                n.depth_db = max(n.target_db, n.depth_db - self.attack_db)
            else:
                n.depth_db = min(n.target_db, n.depth_db + self.release_db)
            n.idle = n.idle + 1 if n.depth_db >= -1e-6 else 0
            # Recompute coefficients for this frame's depth.
            n.b, n.a = peaking_biquad(self.sr, n.freq, n.q, n.depth_db)
            if n.zi is None:
                n.zi = lfilter_zi(n.b, n.a) * 0.0
            if n.idle < 20:  # keep a freed slot around briefly to avoid flapping
                survivors.append(n)
        self.notches = survivors

    # ---- main streaming entry point -------------------------------------
    def process(self, x: np.ndarray) -> np.ndarray:
        out = np.empty_like(x, dtype=np.float64)
        ring = np.zeros(self.win)
        n = len(x)
        idx = 0
        while idx < n:
            block = x[idx : idx + self.hop]
            # Slide the new block into the analysis ring buffer.
            ring = np.roll(ring, -len(block))
            ring[-len(block):] = block
            confirmed = self._detect(ring)
            self._update_notches(confirmed, ring)
            # Apply the notch cascade in the time domain, state-continuous.
            y = block.astype(np.float64)
            for nt in self.notches:
                y, nt.zi = lfilter(nt.b, nt.a, y, zi=nt.zi)
            out[idx : idx + len(block)] = y
            idx += len(block)
        return out

    @property
    def detected_frequencies(self) -> list[float]:
        return list(self._log)
